Add support for registering Amazon Machine Images from EBS snapshots.

gunzy83 opened the pull request:
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME

ec2_ami
##### ANSIBLE VERSION

```
ansible --version
ansible 2.1.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/usr/share/ansible']
```
##### SUMMARY

This pull request implements registration of Amazon Machine Images using an EBS Snapshot instead of an instance. This allows for an AMI to be created from a backup snapshot.

Extra parameters (architecture, kernel_id, virtualization_type, root_device_name) have been added to support the `ec2.register_image` boto method. The difference in block device mapping variable names between `ec2.register_image` and `ec2.create_image` had to be worked around.

Other changes include:
- Ensure that launch_permissions are updated when an AMI with the same name (no `image_id` specified) already exists
- Update register variable names in examples to be more consistent with the purpose of the module

gunzy83 committed:
Add support for registering Amazon Machine Images from EBS snapshots.

gregdek commented:
Thanks @gunzy83. To the current maintainers, @jjshoe please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

jjshoe commented:
@gunzy83 Can you provide a set of steps so this can be tested?

jjshoe commented:
#4075

gunzy83 commented:
Here is an example that will work that demonstrates our use case:

``` yaml
---
- hosts: localhost
  connection: local
  gather_facts: yes

  vars:
    aws_region: ap-southeast-2

  tasks:
    - name: Get Latest Amazon Base AMI
      ec2_ami_find:
        owner: amazon
        region: "{{ aws_region }}"
        name: "amzn-ami-hvm-????.??.?.x86_64-gp2"
        virtualization_type: hvm
        sort: name
        sort_order: descending
        state: available
      register: amis

    - name: Create a copy of the AMI
      ec2_ami:
        name: "myami-{{ ansible_date_time.iso8601 | replace(':','-') }}"
        region: "{{ aws_region }}"
        state: present
        wait: yes
        architecture: "{{ amis.results[0].architecture }}"
        virtualization_type: "{{ amis.results[0].virtualization_type }}"
        root_device_name: "{{ amis.results[0].root_device_name }}"
        device_mapping:
          - device_name: /dev/xvda
            size: 8
            snapshot_id: "{{ amis.results[0].block_device_mapping[amis.results[0].root_device_name].snapshot_id }}"
            delete_on_termination: true
            volume_type: gp2
      register: ami_result

    - name: Print result
      debug:
        msg: "{{ ami_result }}"
```

Change `aws_region` as needed.

Our use case is actually making a copy of our own AMI, not the Amazon one but in this example will work for your testing.

Let me know if you need any clarifications.

gunzy83 commented:
Hi,

With the release of 2.2 I have rebased this against `devel` and updated the commit to target 2.3. Build has failed on an incorrect `version_added` for a new option in `network/ios/ios_config.py` merged in 10 hours ago in #5524.

Please advise.

gregdek commented:
Thanks @gunzy83 for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gunzy83 commented:
ready_for_review

gregdek commented:
Thanks @gunzy83. To the current maintainers, @jjshoe please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]
